sw/device/lib/crypto/drivers/aes.c

```diff
@@ -0,0 +1,158 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/crypto/drivers/aes.h"
+
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/base/mmio.h"
+
+#include "aes_regs.h"  // Generated.
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+/**
+ * Spins until the AES hardware reports a specific status bit.
+ */
+static aes_error_t spin_until(uint32_t bit) {
+  mmio_region_t aes = mmio_region_from_addr(TOP_EARLGREY_AES_BASE_ADDR);
+  while (true) {
+    uint32_t reg = mmio_region_read32(aes, AES_STATUS_REG_OFFSET);
+    if (bitfield_bit32_read(reg, AES_STATUS_ALERT_RECOV_CTRL_UPDATE_ERR_BIT) ||
+        bitfield_bit32_read(reg, AES_STATUS_ALERT_FATAL_FAULT_BIT)) {
+      return kAesInternalError;
+    }
+    if (bitfield_bit32_read(reg, bit)) {
+      return kAesOk;
+    }
+  }
+}
+
+aes_error_t aes_begin(aes_params_t params) {
+  aes_error_t err = spin_until(AES_STATUS_IDLE_BIT);
+  if (err != kAesOk) {
+    return err;
+  }
+
+  mmio_region_t aes = mmio_region_from_addr(TOP_EARLGREY_AES_BASE_ADDR);
+
+  uint32_t ctrl_reg = AES_CTRL_SHADOWED_REG_RESVAL;
+  ctrl_reg = bitfield_field32_write(
+      ctrl_reg, AES_CTRL_SHADOWED_OPERATION_FIELD,
+      params.encrypt ? AES_CTRL_SHADOWED_OPERATION_VALUE_AES_ENC
+                     : AES_CTRL_SHADOWED_OPERATION_VALUE_AES_DEC);
+
+  uint32_t mode_code;
+  switch (params.mode) {
+    case kAesCipherModeEcb:
+      mode_code = AES_CTRL_SHADOWED_MODE_VALUE_AES_ECB;
+      break;
+    case kAesCipherModeCbc:
+      mode_code = AES_CTRL_SHADOWED_MODE_VALUE_AES_CBC;
+      break;
+    case kAesCipherModeCfb:
+      mode_code = AES_CTRL_SHADOWED_MODE_VALUE_AES_CFB;
+      break;
+    case kAesCipherModeOfb:
+      mode_code = AES_CTRL_SHADOWED_MODE_VALUE_AES_OFB;
+      break;
+    case kAesCipherModeCtr:
+      mode_code = AES_CTRL_SHADOWED_MODE_VALUE_AES_CTR;
+      break;
+  }
+  ctrl_reg =
+      bitfield_field32_write(ctrl_reg, AES_CTRL_SHADOWED_MODE_FIELD, mode_code);
+
+  uint32_t len_code;
+  switch (params.key_len) {
+    case kAesKeyLen128:
+      len_code = AES_CTRL_SHADOWED_KEY_LEN_VALUE_AES_128;
+      break;
+    case kAesKeyLen192:
+      len_code = AES_CTRL_SHADOWED_KEY_LEN_VALUE_AES_192;
+      break;
+    case kAesKeyLen256:
+      len_code = AES_CTRL_SHADOWED_KEY_LEN_VALUE_AES_256;
+      break;
+  }
+  ctrl_reg = bitfield_field32_write(ctrl_reg, AES_CTRL_SHADOWED_KEY_LEN_FIELD,
+                                    len_code);
+
+  ctrl_reg = bitfield_bit32_write(
+      ctrl_reg, AES_CTRL_SHADOWED_MANUAL_OPERATION_BIT, false);
+  mmio_region_write32_shadowed(aes, AES_CTRL_SHADOWED_REG_OFFSET, ctrl_reg);
+
+  ptrdiff_t share0 = AES_KEY_SHARE0_0_REG_OFFSET;
```

Comment:
Need to spin until idle before writing the key (writing the control register can cause a reseed operation)